### PR TITLE
Improve compilation of short-circuit operators

### PR DESCRIPTION
**Pull request by lpw25**
The compiler optimises short-circuit operations in if conditions. For example:

```ocaml
let foo x y =
  if (x > 4 && not (x < 10)) || (y > 3 && not (y < 9)) then
    x
  else
    0
```

gets compiled to the following code (from `-dlinear`):

```
camlFoo__foo_1002:
  if x/29[%rax] <=s 9 goto L101
  if x/29[%rax] >=s 21 goto L100
  L101:
  if y/30[%rbx] <=s 7 goto L102
  if y/30[%rbx] >=s 19 goto L100
  L102:
  I/31[%rax] := 1
  return R/0[%rax]
  L100:
  return R/0[%rax]
```

However, these optimisations are not applied for short-circuit operators outside of if conditions. For example:

```ocaml
let bar x y =
  (x > 4 && not (x < 10)) || (y > 3 && not (y < 9))
```

gets compiled to:

```
camlFoo__bar_1005:
  if x/29[%rax] <=s 9 goto L107
  I/31[%rax] := x/29[%rax] <s 21
  I/32[%rax] := I/31[%rax]  * 2 + 1
  I/33[%rdi] := 4
  I/34[%rdi] := I/34[%rdi] - I/32[%rax]
  goto L106
  L107:
  I/35[%rdi] := 1
  L106:
  if I/34[%rdi] ==s 1 goto L105
  I/41[%rax] := 3
  return R/0[%rax]
  L105:
  if y/30[%rbx] <=s 7 goto L104
  I/37[%rax] := y/30[%rbx] <s 19
  I/38[%rbx] := I/37[%rax]  * 2 + 1
  I/39[%rax] := 4
  I/40[%rax] := I/40[%rax] - I/38[%rbx]
  return R/0[%rax]
  L104:
  I/36[%rax] := 1
  return R/0[%rax]
```

This PR tidies up these optimisations and applies them to all short-circuit operations (and `not`). `foo` is compiled as before, but `bar` is now compiled to:

```
camlFoo__bar_1005:
  if x/29[%rax] <=s 9 goto L105
  if x/29[%rax] >=s 21 goto L104
  L105:
  if y/30[%rbx] <=s 7 goto L106
  I/31[%rax] := y/30[%rbx] >=s 19
  I/32[%rax] := I/31[%rax]  * 2 + 1
  return R/0[%rax]
  L106:
  V/33[%rax] := 1
  return R/0[%rax]
  L104:
  V/34[%rax] := 3
  return R/0[%rax]
```

With flambda turned on these optimisations were broken both inside and outside of if conditions. For example, the `foo` function above would produce:

```
camlFoo__foo_7:
  x_10/29[%rdi] := R/0[%rax]
  if x_10/29[%rdi] <=s 9 goto L103
  I/31[%rax] := x_10/29[%rdi] <s 21
  I/32[%rax] := I/31[%rax]  * 2 + 1
  I/33[%rsi] := 4
  I/34[%rsi] := I/34[%rsi] - I/32[%rax]
  goto L102
  L103:
  I/35[%rsi] := 1
  L102:
  if I/34[%rsi] !=s 1 goto L100
  if y_9/30[%rbx] <=s 7 goto L101
  if y_9/30[%rbx] >=s 19 goto L100
  I/37[%rax] := 1
  return R/0[%rax]
  L101:
  I/36[%rax] := 1
  return R/0[%rax]
  L100:
  R/0[%rax] := x_10/29[%rdi]
  return R/0[%rax]
```

with this PR it produces the same code as the non-flambda compiler.

### CODE REVIEW

File: asmcomp/cmmgen.ml
Comment (mshinwell) [inline asmcomp/cmmgen.ml]: I think we could just call this `Then_true_else_false`.  It's not used very much and will save the comment.

**Review by let-def — APPROVED** [review]

File: asmcomp/cmmgen.ml
Comment (mshinwell) [reply asmcomp/cmmgen.ml]: Actually, I don't find the use of the word "approximation" very helpful here.  Describing this directly in terms of how it affects the result of `transl_if` would probably be clearer.

**Review by mshinwell — CHANGES_REQUESTED** [review]

**Comment by mshinwell** [comment]
OK after fixing the approx type and adding a Changes entry.  (I don't know how it passed Travis without a Changes entry.)  I will keep an eye on benchmarks once this has been merged.

**Comment by yallop** [comment]
> I don't know how it passed Travis without a Changes entry

It's due to the problem described by (and fixed in) #1222.

**Comment by lpw25** [comment]
Rebasing brought in a comment from Pierre about debug info on short-circuit operators, which inspired me to make them a bit more precise. I've also renamed the approx type and added a Changes entry.

**Comment by mshinwell** [comment]
@let-def Please read the diff of the most recent changes again if you could (rev 222d2276 to the tip).

**Review by mshinwell — APPROVED** [review]

**Comment by mshinwell** [comment]
The whole looks ok to me, but waiting for the completed second review from @let-def .

**Comment by let-def** [comment]
I reviewed again. It still looks good to me.

**Comment by mshinwell** [comment]
@xavierleroy We've looked at this carefully, are you ok with this being merged?

**Comment by xavierleroy** [comment]
I haven't had time to review the code in details, but this sounds good, and doubly so if it also fixes a missed optimization in flambda.

The one point where I need to be reassured is that simple Boolean-valued expressions such as
```
   let f x = x < 3
```
are still compiled to branchless instructions (e.g. `cmp / set / movzb` on x86) and don't get compiled by accident like
```
  let f x = if x < 3 then true else false
```

**Comment by lpw25** [comment]
> The one point where I need to be reassured is that...

Yeah, the patch keeps track of that case explicitly with the `then_else` type: https://github.com/ocaml/ocaml/pull/1215/files#diff-c3e630ae63d7483207225e1ecb5a232eR265. So that should always be handled correctly.

**Comment by mshinwell** [comment]
Merging, following @lpw25 's reply to @xavierleroy .

**Pull request merged**
